lightly/api/api_workflow_client.py

```diff
@@ -130,7 +132,47 @@ def filenames_on_server(self):
                 get_sample_mappings_by_dataset_id(dataset_id=self.dataset_id, field="fileName")
         return self._filenames_on_server
 
-    def upload_file_with_signed_url(self, file, signed_write_url: str):
-        response = put_request(signed_write_url, data=file)
-        file.close()
+    def upload_file_with_signed_url(self, file: IOBase, signed_write_url: str,
+                                    max_backoff: int = 32, max_retries: int = 5) -> Response:
+        """Uploads a file to a url via a put request.
+
+        Args:
+            file:
+                The file to upload.
+            signed_write_url:
+                The url to upload the file to. As no authorization is used,
+                the url must be a signed write url.
+            max_backoff:
+                Maximal backoff before retrying.
+            max_retries:
+                Maximum number of retries before timing out.
+
+        Returns:
+            The response of the put request, usually a 200 for the success case.
+
+        """
+
```

Comment:
Why not use [this](https://github.com/lightly-ai/lightly/blob/7e68ddf03b898a9e9c00549918366c3af71ab865/lightly/openapi_generated/swagger_client/api_client.py#L343) instead?

Reply:
I tried to use it for several hours, but it does not have a `data` attribute and expects the data to be json-serialisable, as it is designed for REST apis requests.

Reply:
And what about [this](https://docs.python.org/3/library/urllib.request.html#urllib.request.Request)?

Reply:
But then your `signed_write_url` has expired. This is not a problem of urllib but the server who refuses to authorize your request.

Reply:
urllib2 is not part of the standard package, urllib3 does not have a `data` attribute neither. I looked at many SO-sites for how to send an image/file with one of the urllib libraries and did not find any working ones. Oftentimes, it was recommended to use the requests library instead, also see https://stackoverflow.com/questions/2018026/what-are-the-differences-between-the-urllib-urllib2-urllib3-and-requests-modul